### PR TITLE
Draft ADR 28

### PR DESCRIPTION
**Pull request by kilbergr**
## Description
This is the ADR around state machines in our app. There are multiple supporting docs, without which this won't be that useful.
The first is the real time board (linked in ADR), that you can also find here: https://realtimeboard.com/app/board/o9J_kzRjJ2k=/
In the RTB, the current state is on top (and has already been reviewed).
The Activity Diagram of where we would like to be (not completely future-predicting, but a few extras filled out and some aspects changed) is below.
The State diagrams are at the very bottom.

[Here are the code changes](https://github.com/transcom/mymove/pull/982) for an example of what implementing the suggested FSM library would look like. This implements changes on the Move model. To get it working on your branch, you'll need to run `go get github.com/looplab/fsm`.

## Reviewer Notes

This ADR doesn't include a final decision. That's because I was hoping folks would chime in with their opinions about the two options. As far as I can tell, there's not an overwhelming decision to choose either, which makes me lean towards keeping what we already have in place and building it out. But I'd love your feedback.
Please ask any questions.

## Code Review Verification Steps

* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/160067735) for this change

### CODE REVIEW

File: docs/adr/0028-Backend-State-Machine.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+### *Use a State Machine Library*
+
+#### Decided on [Loop Lab Finite State Machine for Go](https://github.com/looplab/fsm)--most robust, highly recommended
+
+* `+` One place in each model where all implementation occurs.
+* `+` Provides hooks for enter/exit states, regardless of transition
+* `+` Provides hooks for code to run before/after transitions (before transition can block a transition)
+* `-` Have to implement throughout (includes adding a new field with serialization/de-serialization functions to each model that uses).
+* `-` Have to change currently existing Status field to be some form of private because we wouldn't want that to be writable outside of the FSM paradigm
```
Comment (jim) [inline docs/adr/0028-Backend-State-Machine.md]: I'm pretty sure that it isn't possible to do this with Pop structs as the private field wouldn't be accessible by Pop itself for loading and saving the model.

File: docs/adr/0028-Backend-State-Machine.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+* **To determine the difficulty of implementation, [we stubbed out implementing on the Move model here](https://github.com/transcom/mymove/tree/rek_fsm-loop). Please use this to form your opinion of the size of this task.**
+
+### *Keep using our own version*
+
+* `+` Have already defined the majority of necessary transitions.
+* `+` Only write what you need
+* `-` Doesn't save when you fire transitions
+* `-` Each transition/state function has to be written independently
+
```
Comment (jim) [inline docs/adr/0028-Backend-State-Machine.md]: I would add:

`-` Legal transitions are not defined declaratively, possibly making the state graph harder to reason about

File: docs/adr/0028-Backend-State-Machine.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+* Chosen Alternative: *[alternative 1]*
+* *[justification. e.g., only alternative, which meets KO criterion decision driver | which resolves force force | ... | comes out best (see below)]*
+* *[consequences. e.g., negative impact on quality attribute, follow-up decisions required, ...]* <!-- optional -->
+
+## Pros and Cons of the Alternatives
+
+### *Use a State Machine Library*
+
+#### Decided on [Loop Lab Finite State Machine for Go](https://github.com/looplab/fsm)--most robust, highly recommended
+
```
Comment (jim) [inline docs/adr/0028-Backend-State-Machine.md]: I would add:

`-` Additional arguments to events are of type `interface{}`

**Comment by jim** [comment]
I lean towards using the library as I think it will force us to be more consistent with our implementation. It's too bad that Pop prevents us from easily making the `Status` field private, but that's a language limitation.

File: docs/adr/0028-Backend-State-Machine.md (hunk 1 above)
Comment (chrisgilmerproj) [reply docs/adr/0028-Backend-State-Machine.md]: I tried to look into modifying the pop serializer and it seemed like too much work at the time.  We're kind of stuck with this and its not great.

**Comment by chrisgilmerproj** [comment]
I also lean towards using the FSM library.  It would be nice to have the callback functionality without having to write our own.

File: docs/adr/0028-Backend-State-Machine.md (hunk 1 above)
Comment (tinyels) [reply docs/adr/0028-Backend-State-Machine.md]: `state_LEAVE_ME_ALONE`

File: docs/adr/0028-Backend-State-Machine.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+* **To determine the difficulty of implementation, [we stubbed out implementing on the Move model here](https://github.com/transcom/mymove/tree/rek_fsm-loop). Please use this to form your opinion of the size of this task.**
+
+### *Keep using our own version*
+
+* `+` Have already defined the majority of necessary transitions.
+* `+` Only write what you need
+* `-` Doesn't save when you fire transitions
+* `-` Each transition/state function has to be written independently
+
+Comparison of implementation
```
[the remark this answers is not in the document]
Comment (tinyels) [reply docs/adr/0028-Backend-State-Machine.md]: yeah, a `GetValidNextActions` method would be really nice. Even better if it was role-aware.

File: docs/adr/0028-Backend-State-Machine.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+* *Keep using our own state machines*
+* *Use a state machine library*
+  * *[Loop Lab Finite State Machine for Go](https://github.com/looplab/fsm)*
+  * *[QOR/transition](https://github.com/qor/transition)*
+  * *[A simple finite state machine for Golang](https://github.com/ryanfaerman/fsm)*
+
+## Decision Outcome
+
+* Chosen Alternative: *[alternative 1]*
```
Comment (ntwyman) [inline docs/adr/0028-Backend-State-Machine.md]: What is Alternative 1? Are the alternatives numbers from 0 or 1?

Can you be more explicit here please.

File: docs/adr/0028-Backend-State-Machine.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+* *[justification. e.g., only alternative, which meets KO criterion decision driver | which resolves force force | ... | comes out best (see below)]*
```
Comment (ntwyman) [inline docs/adr/0028-Backend-State-Machine.md]: Ok looks like this section is boiler plate. Please fill in

File: docs/adr/0028-Backend-State-Machine.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+## Pros and Cons of the Alternatives
+
+### *Use a State Machine Library*
+
+#### Decided on [Loop Lab Finite State Machine for Go](https://github.com/looplab/fsm)--most robust, highly recommended
+
+* `+` One place in each model where all implementation occurs.
+* `+` Provides hooks for enter/exit states, regardless of transition
+* `+` Provides hooks for code to run before/after transitions (before transition can block a transition)
+* `-` Have to implement throughout (includes adding a new field with serialization/de-serialization functions to each model that uses).
```
Comment (ntwyman) [inline docs/adr/0028-Backend-State-Machine.md]: Why? Why can't we do it piecemeal?

File: docs/adr/0028-Backend-State-Machine.md (hunk 1 above)
Comment (ntwyman) [inline docs/adr/0028-Backend-State-Machine.md]: Why

File: docs/adr/0028-Backend-State-Machine.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+* `-` Have to change currently existing Status field to be some form of private because we wouldn't want that to be writable outside of the FSM paradigm
+* `-` Probably heavier than we really need, since our transitions tend to be quite straightforward
+
+* **To determine the difficulty of implementation, [we stubbed out implementing on the Move model here](https://github.com/transcom/mymove/tree/rek_fsm-loop). Please use this to form your opinion of the size of this task.**
```
Comment (ntwyman) [inline docs/adr/0028-Backend-State-Machine.md]: I'm not sure I see how this is an improvement. Frankly, I didn't find the code very illuminating.

File: docs/adr/0028-Backend-State-Machine.md (hunk 1 above)
Comment (kilbergr) [reply docs/adr/0028-Backend-State-Machine.md]: because we wouldn't want that to be writable outside of the FSM paradigm

**Comment by pjdufour-truss** [comment]
Will need to be bumped to ADR 29.  My deepest apologies.

**Comment by jim** [comment]
Closing as this hasn't been updated in >= 6 weeks. Updating the state machine machinery is on hold while we sort out the service object stuff. We may return to it after that.

**Pull request closed without merging**